ydb/core/persqueue/partition.cpp

```diff
@@ -2258,83 +2296,101 @@ void TPartition::ProcessImmediateTx(const NKikimrPQ::TEvProposeTransaction& tx,
+        result = PreProcessRequest(msg.GetDeregisterMessageGroup());
+    } else if (msg.IsSplitMessageGroup()) {
+        result = PreProcessRequest(msg.GetSplitMessageGroup());
+    } else {
+        Y_ABORT_UNLESS(msg.IsOwnership());
+    }
+    return result;
 }
 
-TPartition::EProcessResult TPartition::ProcessUserActionOrTransaction(TMessage& msg,
-                                                                      TEvKeyValue::TEvRequest* request,
-                                                                      const TActorContext& ctx)
+void TPartition::ExecUserActionOrTransaction(TMessage& msg, TEvKeyValue::TEvRequest* request, bool committed)
 {
-    if (!HaveWriteMsg) {
+    Y_ABORT_UNLESS(committed);
+    const auto& ctx = ActorContext();
+    if (!HaveWriteMsg) { //ToDo: !! check it is not set on preprocessing;
         BeginHandleRequests(request, ctx);
         if (!DiskIsFull) {
             BeginProcessWrites(ctx);
             BeginAppendHeadWithNewWrites(ctx);
         }
-
         HaveWriteMsg = true;
     }
 
-    if (WriteCycleSize >= MAX_WRITE_CYCLE_SIZE) {
-        return EProcessResult::Break;
-    }
-
-    auto result = EProcessResult::Continue;
+    bool doReply = true;
     if (msg.IsWrite()) {
-        result = ProcessRequest(msg.GetWrite(), *Parameters, request, ctx);
+        doReply = ExecRequest(msg.GetWrite(), *Parameters, request);
     } else if (msg.IsRegisterMessageGroup()) {
-        result = ProcessRequest(msg.GetRegisterMessageGroup(), *Parameters);
+        ExecRequest(msg.GetRegisterMessageGroup(), *Parameters);
     } else if (msg.IsDeregisterMessageGroup()) {
-        result = ProcessRequest(msg.GetDeregisterMessageGroup(), *Parameters);
+        ExecRequest(msg.GetDeregisterMessageGroup(), *Parameters);
     } else if (msg.IsSplitMessageGroup()) {
-        result = ProcessRequest(msg.GetSplitMessageGroup(), *Parameters);
+        ExecRequest(msg.GetSplitMessageGroup(), *Parameters);
     } else {
         Y_ABORT_UNLESS(msg.IsOwnership());
     }
-
-    if ((result == EProcessResult::Reply) || msg.IsOwnership()) {
+    if (doReply) {
         EmplaceResponse(std::move(msg), ctx);
     }
-
-    return result;
 }
 
-void TPartition::ProcessUserAct(TEvPQ::TEvSetClientInfo& act,
-                                const TActorContext& ctx)
-{
+TPartition::EProcessResult TPartition::PreProcessUserAct(
+        TEvPQ::TEvSetClientInfo& act, const TActorContext&
+) {
     Y_ABORT_UNLESS(!KVWriteInProgress);
 
     const TString& user = act.ClientId;
-    const bool strictCommitOffset = (act.Type == TEvPQ::TEvSetClientInfo::ESCI_OFFSET && act.Strict);
+    if (act.Type == TEvPQ::TEvSetClientInfo::ESCI_OFFSET) {
+        if (TxAffectedClients.contains(user)) {
+            return EProcessResult::Blocked;
+        }
+        SetOffsetAffectedClients.insert(user);
+    }
+    SetOffsetAffectedClients.insert(user);
+    return EProcessResult::ContinueAutoCommit;
```

Comment:
What do ContiueAutoCommit means? And when must be used?

Reply:
No additional TxCommit signal will be expected for this action, it can be processed instantly as soon as we get it from the queue. Used for anything except transactions.